nodeJS/express/controllers.md

```diff
@@ -0,0 +1,460 @@
+      return;
+    }
+
+    res.send(`User found: ${user.name}`);
+  } catch (error) {
+    console.error("Error retrieving user:", error);
+    res.status(500).send("Internal Server Error");
+
+    // or we can call next(error) instead of sending a response here
+    // Using `next(error)` however will only render an error page in the express' default view and respond with the whole html to the client.
+    // So we will need to create a special type of middleware function if we want a different response and we will get to that in a bit.
+  }
+};
+```
+
+However, this is not exactly a clean solution since eventually, we will have to add the same try/catch block to *all* controllers. There is a package called [express-async-handler](https://www.npmjs.com/package/express-async-handler) that you can install to hide this bit. It essentially just chains a `catch` to "catch" any errors in the function.
+
+```javascript
+const asyncHandler = require("express-async-handler");
+
+// Any errors that is thrown in this function will automatically be caught and call the `next` function
+const getUserById = asyncHandler(async (req, res) => {
+  const userId = req.params.id;
+
+  const user = await someDBQueryToGetUser(userId);
+
+  if (!user) {
+    res.status(404).send("User not found");
+    return;
+  }
+
+  res.send(`User found: ${user.name}`);
+});
+```
+
+#### With a middleware
+
+Remember what we said earlier regarding a "special type of middleware"? Let's actually look into that now. There is an error middleware function that handles all errors in our application coming down from other middleware functions and this error middleware function is commonly placed at the very end of our application code, to ensure that it is actually the last middleware function to be executed and to only handle errors bubbling down from other middleware functions before it.
+
+```javascript
+// Every thrown error in the application or the previous middleware function calling `next` with an error as an argument will eventually go to this middleware function
+app.use((err, req, res, next) => {
+  console.error(err);
+  res.status(500).send(err);
+});
+```
+
+However, take note that this is a middleware function that requires *four parameters* that you will need to provide even if they are not used. If you for example exclude one of the parameters, it will not be recognized as an error middleware function. You can try it out yourself ;)
+
+It is the same as the previous middleware functions with three parameters but with one new parameter in a different order:
+
+- `err` - The error object. This is an odd one but it *must* be the first parameter in the callback.
+- `req` - The request object, which represents the incoming HTTP request.
+- `res` - The response object, which represents the HTTP response that will be sent back to the client.
+- `next` - The function that pass the control to the *next* middleware function (We'll get to this later). This is optional.
```

Comment:
we don't need to explain these params again

Reply:
Condensed it to

> It is the same as the previous middleware functions with three parameters but with one new parameter in a different order which is the error object. This is an odd one but the error object *must* be the first parameter in the callback.